File: examples/serving/attention_example.py
import numpy as np
import scipy

np.random.seed(42)

np.set_printoptions(precision=2)

def single_attention(q: np.ndarray, K: np.ndarray, V: np.ndarray, o: np.ndarray, i: int, b: int) -> None:
    """_summary_

    Args:
        q (np.ndarray): query vector
        K (np.ndarray): Key cache matrix
        V (np.ndarray): Value cache matrix
        o (np.ndarray): Output vector (including the prompt)
        i (int): i-th index of the output
        b (int): batch index
    """
    dk_sqrt = len(q) ** 0.5

    A = np.zeros((i))

    for j in range(i):

        # A[j] = scipy.special.softmax(np.dot(q, K[b, j, :])/dk_sqrt)
        A[j] = np.exp(np.dot(q, K[b, j, :])/dk_sqrt)

        tmp = 0.0
        for t in range(i):
            tmp += np.exp(np.dot(q, K[b, t, :])/dk_sqrt)

        A[j] /= tmp
        
        o[b, i, :] += A[j]*V[b, j, :]



def main():
    B = 2   # batch size
    n = 8   # max sequence length
    d = 4   # hidden state size (or number of possible tokens)

    assert B < 3, "With one it is looking better"

    # query, key, value matrices filled with random floats
    Wq = np.random.rand(d,d)
    Wk = np.random.rand(d,d)
    Wv = np.random.rand(d,d)

    # KV cache
    K = np.zeros((B, n, d))
    V = np.zeros((B, n, d))

    # prompt + output lists
    o = np.zeros((B, n, d))
    p_size = 3 # prompt size

    # creating prompts
    for b in range(B):
        for i in range(p_size):
            o[b, i, :] = np.random.rand(d)

    # prompt phase
    for b in range(B):
        for i in range(p_size):
            K[b, i, :] = Wk @ o[b, i, :]
            V[b, i, :] = Wv @ o[b, i, :]

    # autoregressive phase
    for b in range(B):
        for i in range(p_size, n):
            # Computing the q,k,v of the given iteration (and storing KV cache)
            q          = Wq @ o[b, i, :]
            K[b, i, :] = Wk @ o[b, i, :]
            V[b, i, :] = Wv @ o[b, i, :]

            # single header attention of the given iteration for the output o_i
            single_attention(q, K, V, o, i, b)

            # print result
            print(f"b: {b}, i: {i}, o:\n{o[b, :, :]}")


if __name__ == "__main__":
    main()

File: examples/serving/attention_example.cpp
#include <torch/torch.h>
#include <math.h>
#include <iostream>

void single_attention(const torch::Tensor& q, const torch::Tensor& KV_cache, torch::Tensor& o, const size_t i, const size_t b)
{
    const auto dk_sqrt = sqrt(double(q.sizes()[0]));

    auto A = torch::zeros({int64_t(i)});

    for(size_t j = 0; j < i; ++j){
        A[j] = exp(torch::dot(q, KV_cache[b][0][j])/dk_sqrt);

        auto tmp = torch::zeros({1});
        for(size_t t = 0; t < i; ++t){
            tmp += exp(torch::dot(q, KV_cache[b][0][j])/dk_sqrt);
        }

        A[j] /= tmp[0];

        o[b][i] += A[j]*KV_cache[b][1][j];
    }
}

int main(int argc, char *argv[])
{
    torch::manual_seed(42);

    // Variables
    const size_t B = 1; // batch size
    const size_t n = 8; // max sequence length
    const size_t d = 4; // hidden state size (or number of possible tokens)

    // query, key, value matrices filled with random floats
    auto Wq = torch::rand({d,d});
    auto Wk = torch::rand({d,d});
    auto Wv = torch::rand({d,d});
    
    // KV cache
    auto KV_cache = torch::zeros({B, 2, n, d});
    
    // prompt + output lists
    auto o = torch::zeros({B, n, d});
    const size_t p_size = 3; // prompt size (this one will later be flexible)

    // creating prompts
    o.slice(1, 0, p_size) = torch::rand({B, p_size, d});

    // prompt phase
    for(size_t b = 0; b < B; ++b){
        for(size_t i = 0; i < p_size; ++i){
            KV_cache[b][0][i] = torch::matmul(Wk, o[b][i]);
            KV_cache[b][1][i] = torch::matmul(Wv, o[b][i]);
        }
    }

    // autoregressive phase
    for(size_t b = 0; b < B; ++b){
        for(size_t i = p_size; i < n; ++i){
            auto q            = torch::matmul(Wq, o[b][i]);
            KV_cache[b][0][i] = torch::matmul(Wk, o[b][i]);
            KV_cache[b][1][i] = torch::matmul(Wv, o[b][i]);

            single_attention(q, KV_cache, o, i, b);

            std::cout << "b: " << b << ", i: " << i << ", o: " << o << "\n";
        }
    }

    return 0;
}